src/devices/GrovePi/GrovePi.cs

```diff
@@ -0,0 +1,190 @@
+        }
+
+        /// <summary>
+        /// Get the firmware version
+        /// </summary>
+        /// <returns>GroovePi firmware version</returns>
+        public Version GetFirmwareVerion()
+        {
+            WriteCommand(GrovePiCommands.Version, 0, 0, 0);
+            Thread.Sleep(10);
+            var inArray = ReadCommand(GrovePiCommands.Version, 0);
+            return new Version(inArray[1], inArray[2], inArray[2]);
+        }
+
+        /// <summary>
+        /// Write a GrovePi command
+        /// </summary>
+        /// <param name="commands">The GrovePi command</param>
+        /// <param name="pin">The pin to write the command</param>
+        /// <param name="param1">First parameter</param>
+        /// <param name="param2">Second parameter</param>
+        public void WriteCommand(GrovePiCommands commands, GrovePort pin, byte param1, byte param2)
+        {
+            byte[] outArray = new byte[4] { (byte)commands, (byte)(pin), param1, param2 };
+            try
+            {
+                _i2cDevice.Write(outArray.AsSpan());
+            }
+            catch (Exception ex) when (ex is IOException)
+            {
+                // Give it a second try
+                _i2cDevice.Write(outArray.AsSpan());
+            }
+        }
+
+        /// <summary>
+        /// Read data from GrovePi
+        /// </summary>
+        /// <param name="commands">The GrovePi command</param>
+        /// <param name="pin">The pin to read</param>
+        /// <returns></returns>
+        public byte[] ReadCommand(GrovePiCommands commands, GrovePort pin)
+        {
+            int numberBytesToRead = 0;
+            switch (commands)
+            {
+                case GrovePiCommands.DigitalRead:
+                    numberBytesToRead = 1;
+                    break;
+                case GrovePiCommands.AnalogRead:
+                case GrovePiCommands.UltrasonicRead:
+                case GrovePiCommands.LetBarGet:
+                    numberBytesToRead = 3;
+                    break;
+                case GrovePiCommands.Version:
+                    numberBytesToRead = 4;
+                    break;
+                case GrovePiCommands.DhtTemp:
+                    numberBytesToRead = 9;
+                    break;
+                // No other commands are for read
+                default:
+                    return null;
+            }
+            Span<byte> inArray = new Span<byte>(new byte[numberBytesToRead]);
+            _i2cDevice.Read(inArray);
+            return inArray.ToArray();
+        }
+
+        /// <summary>
+        /// Read a digital pin, equivalent of digitalRead on Arduino
+        /// </summary>
+        /// <param name="pin">The GroovePi pin to read</param>
+        /// <returns>Returns the level either High or Low</returns>
+        public PinLevel DigitalRead(GrovePort pin)
+        {
+            WriteCommand(GrovePiCommands.DigitalRead, pin, 0, 0);
+            Thread.Sleep(10);
```

Comment:
why is the sleep needed? if datasheet requires it - should it be part of WriteCommand instead?

Reply:
Yes, it's needed to have the operation done correctly on the GrovePi side. Now, As  implemented a retry method in the write command, I can try without and run tests like this.

Reply:
perhaps add this delay before doing the retry

Reply:
Yes, exactly what I was trying to test. I did a stress test wihout any wait. And after 15 minutes it failed. Will run a new one with the wait before the retry

Reply:
I ran more tests and still failing after 30 minutes with older firmware (1.2.2). I'll implement multiple retries with the wait pattern. Let's see.